File: endpoints/java/ReliableBench.java
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 ZeroDDS Contributors
//
// Producer-latency micro-bench: the async-decoupled writer's BlockingQueue
// enqueue vs. an inline UDP DatagramSocket#send. Prints median ns for both --
// the syscall the enqueue avoids on the producer's hot path. Mirrors
// endpoints/d/reliable_bench.d (SpscRing.enqueue vs. sendTo).

import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.util.Arrays;
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;

import org.zerodds.endpoint.ReliableWire;

public final class ReliableBench {

    private ReliableBench() {}

    public static void main(String[] args) throws Exception {
        final int iters = 20_000;
        byte[] frame = ReliableWire.writeFrame(0, new byte[] {1, 2, 3, 4});

        InetAddress host = InetAddress.getByName("127.0.0.1");
        DatagramSocket sink = new DatagramSocket(0, host); // bound, never read -- a real destination

        // Inline path: a real sendto per sample (kernel transition).
        DatagramSocket sock = new DatagramSocket();
        DatagramPacket pkt = new DatagramPacket(frame, frame.length, host, sink.getLocalPort());
        long[] inlineNs = new long[iters];
        for (int i = 0; i < iters; i++) {
            long t0 = System.nanoTime();
            sock.send(pkt);
            inlineNs[i] = System.nanoTime() - t0;
        }

        // Decoupled path: BlockingQueue enqueue only (no syscall on the producer).
        // A daemon drain thread keeps the queue from saturating, the way
        // AsyncReliableWriter's real drain thread would.
        final BlockingQueue<byte[]> queue = new ArrayBlockingQueue<byte[]>(1024);
        final Thread drain = new Thread(new Runnable() {
            @Override
            public void run() {
                try {
                    while (!Thread.currentThread().isInterrupted()) {
                        queue.take();
                    }
                } catch (InterruptedException e) {
                    // stop
                }
            }
        }, "bench-drain");
        drain.setDaemon(true);
        drain.start();

        long[] queueNs = new long[iters];
        for (int i = 0; i < iters; i++) {
            long t0 = System.nanoTime();
            queue.offer(frame);
            queueNs[i] = System.nanoTime() - t0;
        }
        drain.interrupt();

        Arrays.sort(inlineNs);
        Arrays.sort(queueNs);
        long inlineMed = inlineNs[iters / 2];
        long queueMed = queueNs[iters / 2];
        double ratio = queueMed == 0 ? (double) inlineMed : (double) inlineMed / (double) queueMed;
        System.out.printf(
            "producer latency: queue-enqueue median = %d ns, inline-send median = %d ns (%.0fx)%n",
            queueMed, inlineMed, ratio);

        sock.close();
        sink.close();
    }
}

File: endpoints/ocaml/reliable_bench.ml
(* SPDX-License-Identifier: Apache-2.0 *)
(* Copyright 2026 ZeroDDS Contributors *)
(*
   Producer-latency micro-bench: the async-decoupled writer's Mailbox.put
   (mutex + list-cons + condition-signal, no syscall) vs. an inline UDP
   sendto. Prints median ns/op for both -- the syscall the enqueue avoids on
   the producer's hot path.

   `Unix.gettimeofday` has only microsecond-ish resolution, well above a
   single mutex+cons (tens of ns), so each sample times a whole `batch` of
   operations and divides -- the standard technique for sub-clock-resolution
   micro-benchmarks. *)

open Reliable

let batch = 200
let samples = 500

(** Median ns/op over [samples] batches of [batch] calls to [f]. *)
let median_ns_per_op (f : unit -> unit) : float =
  let ns = Array.make samples 0.0 in
  for s = 0 to samples - 1 do
    let t0 = Unix.gettimeofday () in
    for _ = 1 to batch do
      f ()
    done;
    let elapsed = Unix.gettimeofday () -. t0 in
    ns.(s) <- elapsed *. 1_000_000_000.0 /. float_of_int batch
  done;
  Array.sort compare ns;
  ns.(samples / 2)

let () =
  let frame = reliable_write_frame 0 (Bytes.of_string "\001\002\003\004") in

  (* Inline path: a real sendto per sample (kernel transition). *)
  let sock = Unix.socket Unix.PF_INET Unix.SOCK_DGRAM 0 in
  Unix.bind sock (Unix.ADDR_INET (Unix.inet_addr_loopback, 0));
  let sink = Unix.socket Unix.PF_INET Unix.SOCK_DGRAM 0 in
  Unix.bind sink (Unix.ADDR_INET (Unix.inet_addr_loopback, 0));
  let sink_addr = Unix.getsockname sink in

  let inline_med =
    median_ns_per_op (fun () -> ignore (Unix.sendto sock frame 0 (Bytes.length frame) [] sink_addr))
  in

  (* Decoupled path: Mailbox.put only -- no syscall on the producer side.
     Drain after each batch to bound memory (the real drain thread does this
     concurrently, off the producer's hot path). *)
  let mailbox = Mailbox.create () in
  let ring_med = median_ns_per_op (fun () -> Mailbox.put mailbox frame) in
  ignore (Mailbox.drain_all mailbox);

  Printf.printf
    "producer latency: mailbox-enqueue median = %.1f ns/op, inline-sendto median = %.1f ns/op \
     (%.1fx)\n"
    ring_med inline_med
    (if ring_med <= 0.0 then 0.0 else inline_med /. ring_med);
  Unix.close sock;
  Unix.close sink
